internal/webhook/management_webhook.go

```diff
@@ -55,21 +58,109 @@ func (*ManagementValidator) ValidateCreate(_ context.Context, _ runtime.Object)
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (v *ManagementValidator) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+func (v *ManagementValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
 	const invalidMgmtMsg = "the Management is invalid"
 
-	mgmt, ok := newObj.(*hmcv1alpha1.Management)
+	newMgmt, ok := newObj.(*hmcv1alpha1.Management)
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected Management but got a %T", newObj))
 	}
 
+	oldMgmt, ok := oldObj.(*hmcv1alpha1.Management)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected Management but got a %T", oldObj))
+	}
+
+	if err := checkComponentsRemoval(ctx, v.Client, oldMgmt, newMgmt); err != nil {
+		return admission.Warnings{"Some of the providers cannot be removed"},
+			apierrors.NewInvalid(newMgmt.GroupVersionKind().GroupKind(), newMgmt.Name, field.ErrorList{
+				field.Forbidden(field.NewPath("spec", "providers"), err.Error()),
+			})
+	}
+
+	incompatibleContracts, err := getIncompatibleContracts(ctx, v, newMgmt)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %v", invalidMgmtMsg, err)
+	}
+
+	if incompatibleContracts != "" {
+		return admission.Warnings{"The Management object has incompatible CAPI contract versions in ProviderTemplates"}, fmt.Errorf("%s: %s", invalidMgmtMsg, incompatibleContracts)
+	}
+
+	return nil, nil
+}
+
+func checkComponentsRemoval(ctx context.Context, cl client.Client, oldMgmt, newMgmt *hmcv1alpha1.Management) error {
+	removedComponentNames := []string{}
+	for _, oldComp := range oldMgmt.Spec.Providers {
+		if !slices.ContainsFunc(newMgmt.Spec.Providers, func(newComp hmcv1alpha1.Provider) bool { return oldComp.Name == newComp.Name }) {
+			removedComponentNames = append(removedComponentNames, oldComp.Name)
+		}
+	}
+
+	if len(removedComponentNames) == 0 {
+		return nil
+	}
+
+	removedProvidersSet := make(map[string]struct{})
+	for _, m := range removedComponentNames {
+		c, ok := oldMgmt.Status.Components[m]
+		if !ok {
+			continue
+		}
+
+		// it does not matter if component has been successfully installed
+		prTpl := new(hmcv1alpha1.ProviderTemplate)
+		if err := cl.Get(ctx, client.ObjectKey{Name: c.Template}, prTpl); err != nil {
+			return fmt.Errorf("failed to get ProviderTemplate %s: %w", c.Template, err)
+		}
+
+		for _, pn := range prTpl.Status.Providers {
+			removedProvidersSet[pn] = struct{}{}
+		}
+	}
+
+	if len(removedProvidersSet) == 0 { // sanity
+		return nil
+	}
+
+	mcls := new(hmcv1alpha1.ManagedClusterList)
+	if err := cl.List(ctx, mcls); err != nil {
```

Comment:
we have an indexer to select managedclusters by their templates that I think we should use here

Reply:
as of my understanding, field match works only for a single entry (if no custom selector is implemented), if so it requires multiple lists having multiple template names, so both implementations seem equal to me

Reply:
you should not be limited to just `MatchingFields`, but even if you were, it would still be better to have a separate api call per template with `Limit: 1`

Reply:
took a quick look, the existing indexers related to the `clustertemplate` resource and not the `providertemplates`, so basically the idea is to acquire the provider names, for this purpose the `providertemplates` should be fetched, and then we have to fetch all the `clustertemplates` related to a `managedclusters` object in order to determine whether the provider names are actually in use, and the latter requires to fetch all `managedclusters` objects having no `clustertemplates` preliminary. So I cannot see how I can leverage the indexer in such a case tbh

Reply:
yes, the idea was to identify `clustertemplates` associated with the removed `providertemplates` and then query `managedclusters` just once